man/man1/lgt2md.1
.TH lgt2md 1 "December 16, 2024" "Logtalk 3.86.0" "Logtalk Documentation"

.SH NAME
lgt2md \- Logtalk script for converting XML documenting files into Markdown text files

.SH SYNOPSIS
.B lgt2md
[\fB-d \fIdirectory\fR]
[\fB-i \fIindex\fR]
[\fB-t \fItitle\fR]
[\fB-p \fIprocessor\fR]
.br
.B lgt2md
.B \-h

.SH DESCRIPTION
\fBlgt2md\fR is a Logtalk script for converting all Logtalk XML documenting files in the current directory into Markdown text files. Assumes UTF-8 encoding.

.SH OPTIONS
.TP
.B \-d
.I directory
Output directory for the Markdown text files.
.TP
.B \-i
.I index
Name of the index file.
.TP
.B \-t
.I title
Title to be used in the index file.
.TP
.B \-p
.I processor
XSLT processor.
.TP
.B \-h
Prints a summary of the script options.

.SH REQUIREMENTS
Requires a compatible XSLT processor. See the script help screen for more information.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.dtd
The Logtalk entity DTD file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.dtd
The Logtalk index DTD file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/custom.ent
The Logtalk custom entities file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.xsd
The Logtalk entity XML Schema file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.xsd
The Logtalk index XML Schema file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_entity_to_md.xsl
The Logtalk entity XSLT file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_index_to_md.xsl
The Logtalk index XSLT file.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR
.PP
logtalk_user_setup(1),\ lgt2html(1),\ lgt2xml(1),\ lgt2pdf(1),\ lgt2txt(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2024 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2024 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.87.0 - January ??, 2025
=========================

Logtalk compiler and runtime
----------------------------

* ADDED: New `includes/2`, `object/3`, `protocol/3`, and `category/3`
properties to the `logtalk::loaded_file_property/2` predicate.

Installers and installation scripts
-----------------------------------

* UPDATED: Synced the MacPorts portfile with the version used upstream.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* FIXED: VSCode support for finding predicate implementations when the
definition is found in an included file.

* FIXED: VSCode support for finding predicate references when the call is
found in an included file.

* FIXED: VSCode support for finding predicate declarations when the reference
is found in an included file.

* FIXED: VSCode support for navigating call hierarchies where an exception
would be thrown in some cases.


3.86.0 - December 18, 2024
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Experimental `logtalk::message_prefix_file/6` hook predicate to
the message printing mechanism.

* FIXED: False left-recursion warning when compiling a grammar rule for
a multifile non-terminal that calls a local non-terminal with the same
indicator.

Prolog adapter and integration files
------------------------------------

* UPDATED: Ciao Prolog adapter file to pretend that modules are supported
to allow loading libraries and examples that require file `use_module/1`
directives.

Documentation
-------------

* ADDED: Notes on encrypting applications using XVM.

* FIXED: Multiple Handbook grammar and spelling typos.

* FIXED: Multiple man pages grammar and spelling typos.

Library
-------

* FIXED: The `os::sleep/1` predicate for XSB to also accept floats.

Tools
-----

* CHANGED: The `lgtunit` tool test compilation warning and error messages to
include the file context for better integration with other developer tools.

* CHANGED: The `diagrams` tool now uses a Unicode glyph for the magnifying
lens instead of a PNG image.

* CHANGED: The `diagrams` tool now uses a Graphviz `box` shape for Prolog
modules instead of a `tab`, which is now only used for directories and
libraries.

* REMOVED: Deprecated `lgt2svg.js` script from the `diagrams` tool.

* ADDED: Experimental `diagrams` tool support for the d2 graph language.

* UPDATED: The `diagrams` tool to support graph languages such as d2 where the
node identifiers are not global.

* UPDATED: The `diagrams` tool to also support the PearAI, Windsurf, PyCharm,
and IDEA text editors when generating diagrams with local file links.

* UPDATED: The `tutor` tool to also explain duplicated predicate/non-terminal
reference linter warnings.

* UPDATED: The `tutor` tool to call the new `logtalk::message_prefix_file/6`
hook predicate to support VSCode displaying warnings and errors augmented with
explanations.

* UPDATED: The `help` tool for the new `logtalk::message_prefix_file/6` hook
predicate.

* FIXED: The `diagrams` tool to also generate an `instance/class` caption for
metaclasses.

* FIXED: Bug in the `diagrams` tool where repeated nodes could be created for
external entities.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would miss
"includes" edges.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would show
duplicated "includes" edges.

* FIXED: The `diagrams` tool support for the graph languages to quote atoms
in predicate labels and predicate node contents when required.

* FIXED: The `diagrams` tool exclude options to allow generating external
nodes for the excluded directories, files, libraries, and entities.

* FIXED: Bug in the `lgtdoc` tool when printing a warning on missing entity
directive `info/1` key.

Examples
--------

* ADDED: New `permutations` example providing benchmarks based on list
permutations. Adapted from Prolog code written and contributed by Paul
Tarau.

* UPDATED: The `assumptions` example with another individual example.

Tests
-----

* ADDED: Additional tests for the left-recursion linter warning.

* ADDED: Additional tests for the Prolog standard `round/1` evaluable.

Installers and installation scripts
-----------------------------------

* IMPROVED: The manual installation script and MacPorts Portfile instructions
for defining the `LOGTALKHOME` and `LOGTALKUSER` environment variables.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* UPDATED: VSCode support to use the new `logtalk::message_prefix_file/6`
hook predicate to support displaying warnings and errors augmented with
the `tutor` tool explanations.

* UPDATED: VSCode support to also list the `lgtunit` tool warning and error
messages in the "PROBLEMS" pane.

* FIXED: Compilation errors when loading the VSCode support file using the XSB
backend.


3.85.0 - November 12, 2024
==========================

Logtalk compiler and runtime
----------------------------

* CHANGED: The `logtalk_load_context/2` built-in predicate to type-check
the key argument.


tools/lgtdoc/xml/lgt2md.ps1
#############################################################################
## 
##   XML documenting files to Markdown text files conversion script 
##   Last updated on November 27, 2023
## 
##   This file is part of Logtalk <https://logtalk.org/>  
##   Copyright 2022 Hans N. Beck and Paulo Moura <pmoura@logtalk.org>
##   SPDX-License-Identifier: Apache-2.0
##   
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##   
##       http://www.apache.org/licenses/LICENSE-2.0
##   
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
## 
#############################################################################


#Requires -Version 7.3

[CmdletBinding()]
param(
	[Parameter()]
	[String]$d = $pwd, 
	[String]$i = "index.md", 
	[String]$t = "Documentation index", 
	[Switch]$v,
	[Switch]$h
)

function Write-Script-Version {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve
	Write-Output ($myName + " 2.3")
}

function Get-Logtalkhome {
	if ($null -eq $env:LOGTALKHOME) 
	{
		Write-Output "The environment variable LOGTALKHOME should be defined first, pointing"
		Write-Output "to your Logtalk installation directory!"
		Write-Output "Trying the default locations for the Logtalk installation..."
		
		$DEFAULTPATHS = [string[]](
			"C:\Program Files (x86)\Logtalk",
			"C:\Program Files\Logtalk",
			"%LOCALAPPDATA%\Logtalk"
		)
		
		# Checking all possibilites
		foreach ($DEFAULTPATH in $DEFAULTPATHS) { 
			Write-Output ("Looking for: " + $DEFAULTPATH)
			if (Test-Path $DEFAULTPATH) {
				Write-Output ("... using Logtalk installation found at " + $DEFAULTPATH)
				$env:LOGTALKHOME = $DEFAULTPATH
				break
			}
		}
	}
	# At the end LOGTALKHOME was set already or now is set
}

function Get-Logtalkuser {
	if ($null -eq $env:LOGTALKUSER) {
		Write-Output "After the script completion, you must set the environment variable"
		Write-Output "LOGTALKUSER pointing to %USERPROFILE%\Documents\Logtalk."
		$env:LOGTALKUSER = "%USERPROFILE%\Documents\Logtalk"
	}
	# At the end LOGTALKUSER was set already or now is set
}

function Write-Usage-Help() {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve 

	Write-Output "This script converts all Logtalk XML documenting files in the"
	Write-Output "current directory to Markdown text files"
	Write-Output ""
	Write-Output "Usage:"
	Write-Output ("  " + $myName + " [-d directory] [-i index] [-t title]")
	Write-Output ("  " + $myName + " -v")
	Write-Output ("  " + $myName + " -h")
	Write-Output ""
	Write-Output "Optional arguments:"
	Write-Output ("  -d output directory for the generated files (default is " + $d + ")")
	Write-Output ("  -i name of the index file (default is " + $i + ")")
	Write-Output ("  -t title to be used in the index file (default is `"" + $t + "`")")
	Write-Output "  -v print version"
	Write-Output "  -h help"
	Write-Output ""
}

function Check-Parameters() {

	if (-not(Test-Path $d)) { # cannot be ""
		Write-Output ("The " + $p + " output directory does not exist!")
		Start-Sleep -Seconds 2
		Exit
	}

	if ($v -eq $true) {
		Write-Script-Version
		Exit
	}

	if ($h -eq $true) {
		Write-Usage-Help
		Exit
	}

}

function Create-Index-File() {
	New-Item -Path . -Name $i -ItemType "file" -Force > $null

	Add-Content -Path $i -Value ""
	Add-Content -Path $i -Value ("# " + $t)
	Add-Content -Path $i -Value ""
	Add-Content -Path $i -Value ""

	if (Test-Path "directory_index.xml") {
		Add-Content -Path $i -Value "* [Library index](library_index.md)"
		Add-Content -Path $i -Value "* [Directory index](directory_index.md)"
		Add-Content -Path $i -Value "* [Entity index](entity_index.md)"
		Add-Content -Path $i -Value "* [Predicate index](predicate_index.md)"
	} elseif (Get-ChildItem -Path . -Filter .\*.xml | Select-String -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
		Get-ChildItem -Path . -Filter .\*.xml |
		Foreach-Object {
			if ($_ | Select-String -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
				$entity = ($_.BaseName -replace '_[^_]*$')
				$pars   = ($_.BaseName -replace '.*_')
				Write-Output ("  indexing " + $_.BaseName + ".md")
				if ($pars -gt 0) {
					Add-Content -Path $i -Value ("* [" + $entity + "/" + $pars + "](" + $_.BaseName + ".md)")
				} else {
					Add-Content -Path $i -Value ("* [" + $entity + "](" + $_.BaseName + ".md)")
				}
			}
		}
	}

	$date = Get-Date -Format "yyyy-MM-dd-HH:mm:ss"

	Add-Content -Path $i -Value ""
	Add-Content -Path $i -Value ("Generated on " + $date)
}

###################### here it starts ############################ 

Check-Parameters

Get-Logtalkhome

# Check for existence
if (Test-Path $env:LOGTALKHOME) {
	$output = "Found LOGTALKHOME at: " + $env:LOGTALKHOME
	Write-Output $output
} else {
	Write-Output "... unable to locate Logtalk installation directory!"
	Start-Sleep -Seconds 2
	Exit
}

Get-Logtalkuser

# Check for existence
if (Test-Path $env:LOGTALKUSER) {
	if (!(Test-Path $env:LOGTALKUSER/VERSION.txt)) {
		Write-Output "Cannot find version information in the Logtalk user directory at %LOGTALKUSER%!"
		Write-Output "Creating an up-to-date Logtalk user directory..."
		logtalk_user_setup
	} else {
		$system_version = Get-Content $env:LOGTALKHOME/VERSION.txt
		$user_version = Get-Content $env:LOGTALKUSER/VERSION.txt
		if ($user_version -lt $system_version) {
			Write-Output "Logtalk user directory at %LOGTALKUSER% is outdated: "
			Write-Output "    $user_version < $system_version"
			Write-Output "Creating an up-to-date Logtalk user directory..."
			logtalk_user_setup
		}
	}
} else {
	Write-Output "Cannot find %LOGTALKUSER% directory! Creating a new Logtalk user directory"
	Write-Output "by running the logtalk_user_setup shell script:"
	logtalk_user_setup
}

$entity_xslt =$env:LOGTALKUSER + "\tools\lgtdoc\xml\logtalk_entity_to_md.xsl"
$index_xslt  =$env:LOGTALKUSER + "\tools\lgtdoc\xml\logtalk_index_to_md.xsl"

if (!(Test-Path "logtalk_entity.dtd")) {
	Copy-Item -Path $env:LOGTALKHOME\tools\lgtdoc\xml\logtalk_entity.dtd -Destination .
}

if (!(Test-Path "logtalk_index.dtd")) {
	Copy-Item -Path $env:LOGTALKHOME\tools\lgtdoc\xml\logtalk_index.dtd -Destination .
}

if (!(Test-Path "custom.ent")) {
	Copy-Item -Path $env:LOGTALKUSER\tools\lgtdoc\xml\custom.ent -Destination .
}

if (!(Test-Path "logtalk_entity.xsd")) {
	Copy-Item -Path $env:LOGTALKHOME\tools\lgtdoc\xml\logtalk_entity.xsd -Destination .
}

if (!(Test-Path "logtalk_index.xsd")) {
	Copy-Item -Path $env:LOGTALKHOME\tools\lgtdoc\xml\logtalk_index.xsd -Destination .
}

if (Select-String -Path .\*.xml -Pattern '<logtalk' -CaseSensitive -SimpleMatch -Quiet) {
	Write-Output "Converting XML files to Markdown files..."

	$xslt_settings = New-Object System.Xml.Xsl.XsltSettings
	$xslt_settings.EnableDocumentFunction = 1

	$xml_reader_settings = New-Object System.Xml.XmlReaderSettings
	$xml_reader_settings.DtdProcessing = 1

	$xml_url_resolver = New-Object System.Xml.XmlUrlResolver

	$entity_xslt_object = New-Object System.Xml.Xsl.XslCompiledTransform;
	$entity_xslt_object.Load($entity_xslt, $xslt_settings, $xml_url_resolver)

	$index_xslt_object = New-Object System.Xml.Xsl.XslCompiledTransform;
	$index_xslt_object.Load($index_xslt, $xslt_settings, $xml_url_resolver)

	Get-ChildItem -Path .\*.xml |
	Foreach-Object {
		if (Select-String -Path $_ -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
			Write-Output ("  converting " + $_.Name)
			$file = Join-Path $pwd $_.Name
			$md = Join-Path $d ($_.BaseName + ".md")
			$reader = [System.Xml.XmlReader]::Create($file, $xml_reader_settings)
            $fs = New-Object IO.FileStream $md, 'Append', 'Write', 'Read'
			$writer = New-Object System.IO.StreamWriter($fs)
			$entity_xslt_object.Transform($reader, $null, $writer)
			$writer.Close()
			$reader.Dispose()
		}
	}
	Get-ChildItem -Path . -Filter .\*.xml |
	Foreach-Object {
		if (Select-String -Path $_ -Pattern '<logtalk_index' -CaseSensitive -SimpleMatch -Quiet) {
			Write-Output ("  converting " + $_.Name)
			$file = Join-Path $pwd $_.Name
			$md = Join-Path $d ($_.BaseName + ".md")
			$reader = [System.Xml.XmlReader]::Create($file, $xml_reader_settings)
            $fs = New-Object IO.FileStream $md, 'Append', 'Write', 'Read'
			$writer = New-Object System.IO.StreamWriter($fs)
			$entity_xslt_object.Transform($reader, $null, $writer)
			$writer.Close()
			$reader.Dispose()
		}
	}
	Write-Output "conversion done"
	Write-Output ""
	Write-Output ("generating " + $i + " file...")
	Create-Index-File
	Write-Output ($i + " file generated")
	Write-Output ""
} else {
	Write-Output ""
	Write-Output "No XML files exist in the current directory!"
	Write-Output ""
}

if ($pwd -ne (Join-Path $env:LOGTALKHOME xml)) {
	Remove-Item -Path .\logtalk_entity.dtd
	Remove-Item -Path .\logtalk_entity.xsd
	Remove-Item -Path .\logtalk_index.dtd
	Remove-Item -Path .\logtalk_index.xsd
	Remove-Item -Path .\custom.ent
}
